File: src/main/java/com/zerone/core/ZeroneEngine.java
package com.zerone.core;

import java.util.concurrent.SynchronousQueue;
import java.util.concurrent.ThreadPoolExecutor;
import java.util.concurrent.TimeUnit;

import static java.lang.System.nanoTime;

public class ZeroneEngine {

    private ThreadPoolExecutor threadPoolExecutor;
    private SimulatorTask simulatorTask;

    public Object lock;

    public ZeroneEngine() {
        threadPoolExecutor = new ThreadPoolExecutor(1,
                                                    1,
                                                    0,
                                                    TimeUnit.MILLISECONDS,
                                                    new SynchronousQueue<Runnable>(),
                                                    new ThreadPoolExecutor.DiscardPolicy());
        simulatorTask = new SimulatorTask(60);

        lock = new Object();
    }

    public void setSimulationRate(long fps) {
        simulatorTask.setSimulationRate(fps);
    }

    public void start() {
        threadPoolExecutor.execute(simulatorTask);
    }

    public void stop() {
        simulatorTask.stop();
    }

    class SimulatorTask implements Runnable {

        private float NANO_TO_SEC = 1 / 1000000000.0f;

        private Scene scene;

        private boolean shutdown;

        private long duration;

        public SimulatorTask(long simulationRate) {
            setSimulationRate(simulationRate);

            SceneManager.onSceneLoaded.addSubscriber(new Event.Subscriber<Scene>() {
                @Override
                public void onInvoked(Scene param) {
                    scene = param;
                }
            });
        }

        public void setDuration(long duration) {
            this.duration = duration;
        }

        public void setSimulationRate(long fps) {
            duration = TimeUnit.SECONDS.toNanos(1) / fps;
        }

        public void stop() {
            shutdown = true;
        }

        public void onResume() {

        }

        public void onPause() {
            scene.onPause();
            scene.onDestroy();
        }

        @Override
        public void run() {
            onResume();

            long startTime = nanoTime();
            long elapseTime = 0L;

            while (true) {
                Time.setDeltaTime((nanoTime() - startTime) * NANO_TO_SEC);
                startTime = nanoTime();

                synchronized (lock) {
                    // TODO: Use input queue
                    scene.update();

                    lock.notifyAll();
                }

                elapseTime = nanoTime() - startTime;

                if (elapseTime < duration) {
                    try {
                        Thread.sleep(TimeUnit.NANOSECONDS.toMillis(duration - elapseTime));
                    } catch (InterruptedException ignored) {

                    }
                }

                if (shutdown)
                    break;
            }

            shutdown = false;

            onPause();
        }
    }
}

File: src/main/kotlin/com/zerone/core/ZeroneEngine.kt
package com.zerone.core

import java.lang.System.nanoTime
import java.util.concurrent.SynchronousQueue
import java.util.concurrent.ThreadPoolExecutor
import java.util.concurrent.TimeUnit

class ZeroneEngine {

    private val threadPoolExecutor = ThreadPoolExecutor(
            1,
            1,
            0,
            TimeUnit.MILLISECONDS,
            SynchronousQueue(),
            ThreadPoolExecutor.DiscardPolicy()
    )
    private val simulatorTask: SimulatorTask

    var lock: Object

    init {
        simulatorTask = SimulatorTask(60)

        lock = Object()
    }

    fun setSimulationRate(simulationRate: Long) {
        simulatorTask.setSimulationRate(simulationRate)
    }

    fun start() {
        threadPoolExecutor.execute(simulatorTask)
    }

    fun stop() {
        simulatorTask.stop()
    }

    internal inner class SimulatorTask(simulationRate: Long) : Runnable {

        private val NANO_TO_SEC = 1 / 1000000000.0f

        private var scene: Scene? = null

        private var shutdown: Boolean = false

        private var simulationInterval: Long = 0

        init {
            setSimulationRate(simulationRate)
            SceneManager.onSceneLoad.addSubscriber(object : Event.Subscriber<Scene> {
                override fun onEvent(param: Scene) {
                    scene = param
                }
            })
        }

        fun setSimulationRate(simulationRate: Long) {
            simulationInterval = TimeUnit.SECONDS.toNanos(1) / simulationRate
        }

        fun stop() {
            shutdown = true
        }

        fun onResume() {
            // TODO: Validating scene is not null
        }

        fun onPause() {
            scene?.let {
                it.onPause()
                it.onDestroy()
            }
        }

        override fun run() {
            onResume()

            var startTime = nanoTime()
            var elapseTime = 0L

            while (true) {
                Time.deltaTime = (nanoTime() - startTime) * NANO_TO_SEC
                startTime = nanoTime()

                synchronized(lock) {
                    // TODO: Use input queue
                    scene?.update()

                    lock.notifyAll()
                }

                elapseTime = nanoTime() - startTime

                if (elapseTime < simulationInterval) {
                    try {
                        Thread.sleep(TimeUnit.NANOSECONDS.toMillis(simulationInterval - elapseTime))
                    } catch (ignored: InterruptedException) {

                    }
                }

                if (shutdown)
                    break
            }

            shutdown = false

            onPause()
        }
    }
}
